### ISSUE TITLE
React Native Support

### ISSUE THREAD

**Issue by robertpitt** [morrys/react-relay-offline#1]
Hey Lorenzo,

Amazing work so far, I have been thinking about an offline mode in our react native application for a little while now and during the design that your library takes is very similar to what we came up with.

We have been doing some interesting stuff using this approach, such as seeding data into the store via `fetchQuery` and doing some post store updates to simulate root queries, allowing us to perform initial sync and then be able to function completely offline by optimistically adding root nodes to allow offline mode for seeded data rather than just a rolling cache.

I haven't run this project as of yet but I have read over the code, I am just wondering about your choice for using `idb` and it's a potential incompatibility with react-native.

**Comment by morrys** [comment]
Hi Robert,
at the moment to be able to use this library in react-native or to customize the persistence modalities it is necessary to create the offline storage passing the parameters of persistOptions foreseen by the redux-offline library.

Example of using the default storage of the redux-offline library
```
const storeOffline = OfflineStore (network, {});
```

As soon as I have time I will try to better document and improve its configurability.

Thanks anyway for your interest.

**Comment by sibelius** [comment]
I think where to store the data should be customizable

on web you can use localstorage or anything else
on react-native you can use asyncstorage or realm

**Comment by morrys** [comment]
At the moment you can already customize where to save your data. To do this you must create the offline store by passing the persistence options provided by the redux-offline library:

```
const persistOptions = {
     storage: customStorage (),
}
const storeOffline = OfflineStore (network, persistOptions);
```

As default, redux-offline uses the localStorage for the web and the AsyncStorage for react-native.

In the first version of this library I preferred to make an implementation available for the indexeddb as it was not expected. In future versions I could outsource it to reduce the number of dependencies.

It would be useful to have a comparison of the functioning of this library in react-native since at the moment it's not possible to me to verify it.

Thank you

**Comment by robertpitt** [comment]
Hey,

I have created a react-native setup that attempts to initialize your library with react native, you can take a look at the repo here but it's nothing significant.

https://github.com/robertpitt/react-relay-offline-native-example

The first issue I found is that if LoadingComponent is emitted, [src/QueryRendererOffline.tsx#L41](https://github.com/morrys/react-relay-offline/blob/master/src/QueryRendererOffline.tsx#L41) throws because it's trying to render text in a div tag, however, 1) HTML is not available in react-native, 2) text rendered to the screen needs to be wrapped in a `<Text>` component.

The second issue I have found is that with the [Example](https://github.com/robertpitt/react-relay-offline-native-example) noted above, the `QueryRenderer.render` is only being called once, with `{props: null, error: null, retry: null}`

Adding a log in the fetchQuery shows it's not being called.

**Comment by morrys** [comment]
Hi Robert, 
first of all I would like to thank you for the example project, moreover I have just released version 0.0.3 which solves the indexeddb issue.

In order to make your project work, in addition to changing the library version, you will also need to change the creation of the offline memory and the environment in the following way:

```
const storeOffline = OfflineStore (network, {});

export const environment = new Environment ({network, store, dataFrom: "CACHE_FIRST"}, storeOffline);
```

Waiting for your reply
Lorenzo

**Comment by robertpitt** [comment]
![](https://i.imgur.com/QoUwZhn.jpg)

**Comment by morrys** [comment]
Have you changed the creation of the offline store and the environment?

**Comment by robertpitt** [comment]
Yea, here's the local changes that I have.
```diff
@@ -31,11 +24,11 @@ export default network;
 /**
  * Store
  */
-const storeOffline = OfflineStore(network);
+const storeOffline = OfflineStore(network, {});
 const source = new RecordSource(storeOffline);
 export const store = new Store(storeOffline, source);
 
 /**
  * Environment
  */
-export const environment = new Environment({ network, store }, storeOffline);
\ No newline at end of file
+export const environment = new Environment({ network, store, dataFrom: 'CACHE_FIRST' }, storeOffline);
\ No newline at end of file
```

It works when I am in debug mode, I expect it's switching the runtime environment to the one in the debugger rather than the one on the device, the debugger has indexedDB in scope so I think in order to replicate you need to disable debugging mode.

**Comment by morrys** [comment]
Tomorrow, i hope to release a new version where I will set localStorage (for the web) and AsyncStorage (for react native) as default. Thank you

**Comment by robertpitt** [comment]
Amazing.

here's a quick video of the issue where the query renderer doesn't re-render when the store is populated, it seems to be a "cold start" issue that only shows up with first run (no-cache)

https://www.youtube.com/watch?v=qABN08xc60s

**Comment by morrys** [comment]
Hi Robert,
in the new version I have removed the dependency from the indexeddb. Regarding the problem you highlighted in the youtube video I verified that it is a problem of the redux-offline library, which notifies the real state of the network after its rehydration. In the next days I will analyze an alternative solution or I will find a workaround.

**Comment by morrys** [comment]
Hi Robert,
in the new version (0.1.1) i fixed the notification of network state.
Now i think that library is compatible with react-native.

Waiting for your reply in order to close the issue
Lorenzo

**Comment by robertpitt** [comment]
Thanks for the update, Ill test over the next day or two and get back to you with the results.

Great work as well.

**Comment by morrys** [comment]
hi Robert, have you got to try the update?

**Comment by robertpitt** [comment]
Hey @morrys, sorry for the late reply, been quite busy over the past week with work.

So I have just upgraded to `v0.2.0`, which went fairly ok, I have increased the sample code to at least use a fragment container.

Branch Here: https://github.com/robertpitt/react-relay-offline-native-example/tree/feature/fragment-container-v0.2.0, this branch contains just some code that allows me to test the combination of the QueryRenderer, fragments, refetchContainer and pull to refresh.

Things seem to work as expected, there are some areas that I would like to highlight.

1) I would like to see a clearer separation between the redux store and relay store, or maybe internalize the redux store within the environment, so the setup process is more inlined with the typical relay setup.

Questions:
Does this solution still require the need for `PersistGate` to make sure the hydrate is finished before the QueryRenderers's or is that taken care of by the hydrate publishing to the relay store (causing the re-render)

How are mutations handled, I haven't had time to test them but will hopefully look at those within the week, the project I am working on requires a reliable mechanism for mutations to be guaranteed execution, and protection from unexpected errors and application crashes.
